### ISSUE TITLE
transplant error

### ISSUE THREAD

**Issue by GongYanfu** [rokhmatf/robot-vn#1]
I want to know if you have successfully transplanted the model trained by the original author to a real robot? The robot I am using now is also Turtlebot2, which is the same as the robot used by the original author, but after I install the dependencies on the robot and download the trained model, errors will occur when running the two Evaluate files provided by the author. How has your model changed from the original model?

**Comment by rokhmatf** [comment]
I still haven't tried it on real robots, so far I still understand the principles of the paper and conduct experiments through simulations, but later I plan to experiment on real robot.  maybe the middle of this month I will start to prepare for experiments on real robot.  May I know what errors you have encountered in your experiment?

**Comment by rokhmatf** [comment]
Is there any error documentation that you have?, maybe I can help

**Comment by GongYanfu** [comment]
I'll take screentshots later and send it here.
Can the author's code run smoothly on your server? Especially the code about training. I find that you changed the author's code, why did you change it?

**Comment by GongYanfu** [comment]
My current situation: I successfully installed dependencies on the server and successfully ran two Evaluate files on my server. Then I wanted to transplant the author's trained model to a real robot, but the transplant process was not smooth. The dependencies are now installed on the robot, but errors will be reported as soon as any one of two Evaluates runs.
Actually I really hope the original author can answer our questions, but he hasn't responded to messages for a long time.

**Comment by rokhmatf** [comment]
yes the original code can run well on me during training in simulation. the changes in `trainer.py` are only to increase or decrease the `max_time_steps` variable.
maybe if the problem you encounter is very urgent, you can try to ask again about the issue

**Comment by GongYanfu** [comment]
it is what happenedf when I executed the command：python evaluate_dmhouse.py dmhouse --num-episodes 100
But in fact deepmind_lab.so exists in the corresponding path.
![image](https://user-images.githubusercontent.com/46698455/167242840-a89b1c4f-6cd8-485b-bb9d-6dfb9f89ff9c.png)
![image](https://user-images.githubusercontent.com/46698455/167242762-b271aa24-b370-4a2d-a6e5-494a9e78b2d5.png)

**Comment by GongYanfu** [comment]
its the error when I run command：python evaluate_turtlebot.py turtlebot --num-episodes 100
![1651907054(1)](https://user-images.githubusercontent.com/46698455/167243045-d01be795-c10a-4660-9f14-fff696232a26.jpg)
![image](https://user-images.githubusercontent.com/46698455/167243140-ad3cd89f-bd48-4274-9183-81ad2dacc2d0.png)

**Comment by rokhmatf** [comment]
have you followed all the steps written on GitHub?
see from the form of the terminal, do you use some kind of jetson?

**Comment by GongYanfu** [comment]
yes, I followed the steps provided by the author
yes , the Jetson AGX Xavier is used on turtlebot.

**Comment by rokhmatf** [comment]
Have you tried the Colab provided by the author? I think you should try that first.
then for problems with jetson it seems that there are some libraries that are not suitable if installed on jetson.

**Comment by GongYanfu** [comment]
ok， thanks for your suggestion， I will try

**Comment by rokhmatf** [comment]
yes you're welcome, it looks like dmhouse and robot-visual-navigation can run on jetson as the author said in the issue on the original source, but I've never tried it.  good luck.

**Comment by GongYanfu** [comment]
now there is still a question：Is dmhouse needed on a real robot？
according to the paper， my understanding is that dmhouse is used for model pre-training and it is not needed on turtlebot。

**Comment by rokhmatf** [comment]
yes you are right, the robot is only needed to retrieve data and can be trained first to get pre-trained, or can do training directly on the robot.  dmhouse only deals with simulation.

**Comment by GongYanfu** [comment]
Did you transplant the model to the real robot successfully? I haven't worked with robots for two weeks, and I'm taking exams recently

**Comment by GongYanfu** [comment]
Can you give some advice on understanding the code? Especially the code inside the python folder for virtual environment pre-training? I'm an inter-professional student, I've read the code once or feel it's difficult, and I don't understand many parts? Especially the part about the neural network in the A2CAT-VN system given by the author. thanks

**Comment by GongYanfu** [comment]
Regarding the observation image and the target image, where is the code in the neural network that matches the two? Or how to judge that the target point has been reached?
Thank you very much if you can answer my questions.I am the only one in the research group doing deep reinforcement learning.

**Comment by rokhmatf** [comment]
Sorry, the previous message was missed. for experiments with real robots I will postpone it because the robot battery is gone and still looking for a replacement.

**Comment by rokhmatf** [comment]
Basically in code, you just need to run `python train.py <trainer>` in the python folder. as far as I understand, the author for the simulation using the `dmhouse` trainer, so only need to run `python train.py dmhouse`, for hyperparameter settings you can find in the `trainer.py` file then look for a trainer named `dmhouse`. you can find the network model in `model.py` in the `VisualNavigationModel` class

**Comment by rokhmatf** [comment]
for the training process, need the help of `deep-rl` library, if you want to change the total or part of the network, I think you have to change `deep-rl`. to change it, you can access the library on github author(https://github.com/jkulhanek/deep-rl-pytorch) and you can `make` it into a library. at this stage I'm still trying to understand more deeply, if you make any progress or find something you can share with me. thanks

**Comment by GongYanfu** [comment]
ok. thank you very much. but can you tell me or do you know how does the neural network model determine whether the goal is reached?
I dont understand how does the neural network model determine whether the goal is reached?
In image_collection_environment.py I found the following:
![image](https://user-images.githubusercontent.com/46698455/172171425-f643e953-20da-416a-8eab-d819170b4316.png)

But I dont understand it.

**Comment by GongYanfu** [comment]
It seems that the steps given by the author in the ros file home are not enough to make the robot move.
I think some other nodes need to be started here.  I'm tring to find them. sad

**Comment by GongYanfu** [comment]
Do you know if the target image is a color image, a depth image, or something else?

**Comment by GongYanfu** [comment]
Hi, friend. I have successfully deployed the code on the robot, and the robot has moved, the author's code needs some adjustments on the real robot, mainly because the author made some changes but the effect is not very good in a slightly complex environment and the code I still can't understand some places in the book. It seems that the layout of the environment (virtual environment and real office environment) used by the author is that objects are placed next to the wall, and no objects are placed in the middle of the house. This environment seems to be relatively simple.

**Comment by rokhmatf** [comment]
Sorry for the late reply, and I just realized I saw the email just now. This month I focus on exams in my class.

for `image_collection_environment.py` I didn't read it much and didn't change it I left it as is. so I don't really understand that part. I will try to understand it after this.

for ROS on robots, it looks like the turtlebot package should be added, now I haven't tried it because I'm still getting busy. I'm glad to hear that you can experiment.

for the target image, it seems to use the usual rgb image. The depth image is only needed when taking the dataset to be trained. The results of the training will be applied to the robot.

**Comment by rokhmatf** [comment]
For now I'm focusing on developing simulations in more complex environments, and I agree with you that the author's methods are good in simple environments.

**Comment by rokhmatf** [comment]
to be applicable to a more complex environment, it seems to have to make modifications to the program and add some new features.

may i know what you added so that your robot can run?

**Comment by GongYanfu** [comment]
ok I'll tell them to you.
May I ask which communication APP do you usually use? Can we communicate in another way?



---- Replied Message ----
| From | Rokhmat ***@***.***> |
| Date | 06/21/2022 12:05 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [rokhmatf/robot-vn] transplant error (Issue #1) |

to be applicable to a more complex environment, it seems to have to make modifications to the program and add some new features.

may i know what you added so that your robot can run?

**Comment by GongYanfu** [comment]
Do you understand the process of fine tuning? The process how to capture a real-world data set and use it to train the model? I still don't understand the problem.
Do you understand how to use the map_collector pack in the ROS folder?



---- Replied Message ----
| From | Rokhmat ***@***.***> |
| Date | 06/21/2022 12:01 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [rokhmatf/robot-vn] transplant error (Issue #1) |

I'm sorry I can only reply, and I just realized I saw the email just now. This month I focus on exams in my class.

for image_collection_environment.py I didn't read it much and didn't change it I left it as is. so I don't really understand that part. I will try to understand it after this.

for ROS on robots, it looks like the turtlebot package should be added, now I haven't tried it because I'm still getting busy. I'm glad to hear that you can experiment.

for the target image, it seems to use the usual rgb image. The depth image is only needed when taking the dataset to be trained. The results of the training will be applied to the robot.

**Comment by rokhmatf** [comment]
usually i use whatsapp, do you use that?

**Comment by rokhmatf** [comment]
what is meant by fine tuning seems to be training on real-world datasets with previously obtained weights using data from simulations or from previous real-world datasets.

**Comment by GongYanfu** [comment]
Currently I have two challenges:
1. Have you trained other experiments on turtlebot, such as turtlebot-unreal, turtlebot-a2c, turtlebot uses the network model in the paper, but using the author's dataset, the reward of the experimental turtle is lower than that of turtle-a2c, The experimental result about turtlebot is not good.
![image](https://user-images.githubusercontent.com/46698455/176804327-445626a0-f874-4230-b181-0e8ed4b0a8ef.png)

3. How to get the dataset using our  turtlebot, I think that to get the dataset through the robot in the real environment, I need to use the map_collector function package in the ros folder, but I don't know how to use it, and the author does not mention how to use it. I saw that there is not only image information in the dataset, but also some coordinate information in the raster map, which is used to judge whether the target point is reached.

**Comment by GongYanfu** [comment]
I use the WeChat,  I cannt get whatsapp. Can you tel me your email address you use usually. mine is 15228995838@163.com.

**Comment by rokhmatf** [comment]
I've never tried training with a turtlebot, the past few days there have been network problems in my lab. when it recovers i will give it a try.

yes I agree with your opinion, I just now bought a jetson to conduct experiments on the robot, hopefully there will be no problems during preparation.

okay we can communicate via email, my email is always a notification if there is an incoming message. my email is rokhmat1998@gmail.com
